src/Microsoft.ML/Runtime/Internal/Tools/GeneratorUtils.cs

```diff
@@ -0,0 +1,467 @@
+        /// The full name of C would be X.Y.A+B+C. This method will generate the name "ABC" from it. In case
+        /// A is generic with one generic type, then the full name of typeof(A&lt;float&gt;.B.C) would be X.Y.A`1+B+C[[System.Single]].
+        /// In this case, this method will generate the name "ASingleBC".
+        /// </summary>
+        /// <param name="typesSymbolTable">A dictionary containing the names of the classes already generated.
+        /// This parameter is only used to ensure that the newly generated name is unique.</param>
+        /// <param name="type">The type for which to generate the new name.</param>
+        /// <param name="currentNamespace">The namespace prefix to the new name.</param>
+        /// <returns>A unique name derived from the given type and namespace.</returns>
+        public static string GetSymbolFromType(Dictionary<string, string> typesSymbolTable, Type type, string currentNamespace)
+        {
+            var fullTypeName = type.FullName;
+            string name = currentNamespace != "" ? currentNamespace + '.' : "";
+
+            int bracketIndex = fullTypeName.IndexOf('[');
+            Type[] genericTypes = null;
+            if (type.IsGenericType)
+                genericTypes = type.GetGenericArguments();
+            if (bracketIndex > 0)
+            {
+                Contracts.AssertValue(genericTypes);
+                fullTypeName = fullTypeName.Substring(0, bracketIndex);
+            }
+
+            // When the type is nested, the names of the outer types are concatenated with a '+'.
+            var nestedNames = fullTypeName.Split('+');
+            var baseName = nestedNames[0];
+
+            // We currently only handle generic types in the outer most class, support for generic inner classes
+            // can be added if needed.
+            int backTickIndex = baseName.LastIndexOf('`');
+            int dotIndex = baseName.LastIndexOf('.');
+            Contracts.Assert(dotIndex >= 0);
+            if (backTickIndex < 0)
+                name += baseName.Substring(dotIndex + 1);
+            else
+            {
+                name += baseName.Substring(dotIndex + 1, backTickIndex - dotIndex - 1);
+                Contracts.AssertValue(genericTypes);
+                if (genericTypes != null)
+                {
+                    foreach (var genType in genericTypes)
+                    {
+                        var splitNames = genType.FullName.Split('+');
+                        if (splitNames[0].LastIndexOf('.') >= 0)
+                            splitNames[0] = splitNames[0].Substring(splitNames[0].LastIndexOf('.') + 1);
+                        name += string.Join("", splitNames);
+                    }
+                }
+            }
+
+            for (int i = 1; i < nestedNames.Length; i++)
+                name += nestedNames[i];
+
+            Contracts.Assert(typesSymbolTable.Select(kvp => kvp.Value).All(str => string.Compare(str, name) != 0));
```

Comment:
>string.Compare(str, name) != 0 [](start = 82, length = 30)

That's a bit odd. I guess whoever originally wrote this really like `strcmp`. :D `string.Compare(str, name) != 0)` might be preferred as `str == name`, or some sort of ordinal equality.